frontends/smvparser.y

```diff
@@ -511,33 +665,62 @@ basic_expr: constant {
                   smt::Term res = enc.solver_->make_term(smt::Mod, a->getTerm(), b->getTerm());
                   $$ = new SMVnode(res);
               }else{
-                  smt::Term res = enc.solver_->make_term(smt::BVSmod, a->getTerm(), b->getTerm());
-                  $$ = new SMVnode(res);
+                  SMVnode::BVtype bvs_a = a->getBVType();
+                  SMVnode::BVtype bvs_b = b->getBVType();
+                  if (bvs_a == bvs_b == SMVnode::Unsigned){
+                    smt::Term res = enc.solver_->make_term(smt::BVUrem, a->getTerm(), b->getTerm());
+                    $$ = new SMVnode(res,SMVnode::Unsigned);
+                  } else if (bvs_a == bvs_b == SMVnode::Signed){
+                    smt::Term res = enc.solver_->make_term(smt::BVSmod, a->getTerm(), b->getTerm());
+                    $$ = new SMVnode(res,SMVnode::Signed);
+                  } else{
+                    throw CosaException ("Unsigned/Signed unmatch");
+                  } 
               }
             }
             | basic_expr OP_SHIFTR basic_expr{
               SMVnode *a = $1;
               SMVnode *b = $3;
               smt::Term res = enc.solver_->make_term(smt::BVLshr, a->getTerm(), b->getTerm());
-              $$ = new SMVnode(res);
+              $$ = new SMVnode(res,a->getBVType());
             }
             | basic_expr OP_SHIFTL basic_expr{
               SMVnode *a = $1;
               SMVnode *b = $3;
               smt::Term res = enc.solver_->make_term(smt::BVShl, a->getTerm(), b->getTerm());
-              $$ = new SMVnode(res);
+              $$ = new SMVnode(res,a->getBVType());
             }
             | basic_expr sizev {
               throw CosaException("No word1");
             }
             | basic_expr OP_CON basic_expr  {
               SMVnode *a = $1;
               SMVnode *b = $3;
-              smt::Term res = enc.solver_->make_term(smt::Concat, a->getTerm(), b->getTerm());
-              $$ = new SMVnode(res);
+              SMVnode::BVtype bvs_a = a->getBVType();
+              SMVnode::BVtype bvs_b = b->getBVType();
+              if(bvs_a != bvs_b){
+                throw CosaException("Unsigned/Signed unmatch");
+              } else{
+                smt::Term res = enc.solver_->make_term(smt::Concat, a->getTerm(), b->getTerm());
+                $$ = new SMVnode(res,SMVnode::Unsigned);
+              }               
+            }
+            | basic_expr "[" integer_val "]" {
+                throw CosaException("No index Subscript");
             }
             | basic_expr "[" basic_expr ":" basic_expr "]"{
-              throw CosaException("No word1");
+                SMVnode *a = $1;
+                SMVnode *b = $3;
+                SMVnode *c = $5;
+                SMVnode::BVtype bvs_a = a->getBVType();
+                SMVnode::BVtype bvs_b = b->getBVType();
```

Comment:
Why is this `getBVType()` when we would expect `b` and `c` to be integers?

Reply:
I want to do type checking here, so the type of the first expression should be word

Reply:
Hmm not quite. So I agree that the first expression should be a word, so that makes sense that it's type BV. But the second two expressions should be integers, not bit-vectors. So why is it `SMVNode::BVtype bv_b = b->getBVType();` I see that you check if it's an integer below, but I'm confused why it's called a `BVType` in the first place. It seems like it should just be a "Type" and then you check whether it's an integer or not. Do you know what I mean?